### ISSUE TITLE
RenderImage() returns the wrong image for MultiFrame DicomImage

### ISSUE THREAD

**Issue by CDarbonne** [fo-dicom/fo-dicom#805]
### Expected behavior
Using a parallel loop to call `RenderImage()` on the set of frameIndexes for a MultiFrame DicomItem causes `RenderImage()` to return the correct image for the frameIndex passed

### Actual behavior
When a parallel loop is used to call `RenderImage()` on a set of frameIndexes, the frames will often have some misplaced images in the set. 

### Steps to reproduce the behavior
Use a parallel loop to call `RenderImage()` on a set of frameIndexes. To test that it was `RenderImage()` causing the problem, I burned the frameIndex used to retrieve each image into the image as soon as it was returned. While the frameIndexes were in order, the actual images were obviously not correctly ordered, leading me to believe that `RenderImage()` is returning the wrong image sometimes when used in parallel.

```
Parallel.For(0, _dicomImage.NumberOfFrames,
    frameIndex =>
    {
        var frame = CreateBitmap(frameIndex);
        
        lock (this._cachedImages)
        {
            this._cachedImages[frameIndex] = frame;
        }
    });
```

```
private BitmapSource CreateBitmap(int frameIndex)
{
    using (var result = this._dicomImage?.RenderImage(frameIndex))
    {
        var resizedResult = result.AsWriteableBitmap().Resize((Int32)ImageWidth, (Int32)ImageHeight, WriteableBitmapExtensions.Interpolation.Bilinear);

        DrawText(resizedResult, (frameIndex + 1).ToString(), 25);
    }
}

```
To be clear, this exact code runs perfectly if a regular for loop is used in place of the parallel loop.

### fo-dicom version and OS/platform
fo-dicom 4.0.0 / WPF

**Comment by mia-san** [comment]
while I don't have multiframe image handy now, does following work?

```
private static BitmapSource CreateBitmap(DicomDataset dataset, int frameIndex)
{
    var image = new DicomImage(dataset, frameIndex);
    using (var iimage = image.RenderImage(frameIndex))
    {
        return iimage.AsWriteableBitmap();
    }
}
```

**Comment by gofal** [comment]
yes, I see the issue. For performance/caching reasons there is a class-field "_pixels" holding the pixel data. this field is loaded whenever CurrentIndex changes. While setting the property CurrentIndex is within a lock and the loading of _pixels itself is within a lock, the setting of these two properties and then later the usage of the _pixels to generate a BitmapSource out from it are not locked. 

With in the RenderImage method here the class-field _pixels is set: https://github.com/fo-dicom/fo-dicom/blob/fe83639b4c3bf0d5acc557e262578e78fcd9e3c3/DICOM/Imaging/DicomImage.cs#L228
and here it is then used: https://github.com/fo-dicom/fo-dicom/blob/fe83639b4c3bf0d5acc557e262578e78fcd9e3c3/DICOM/Imaging/DicomImage.cs#L237.

To solve it you could declare a local method-variable of type IPixelData and in line 228 you assign it to the class property AND to your local variable within the lock. and then in line 237 you use the local variable instead of the class field (that might have been changed meanwhile).

Please give it a try.

**Comment by CDarbonne** [comment]
Yes @gofal, that worked! Thanks for the help!!

I did notice something else though while using `RenderImage()`. Sometimes _pixels and new local-method pixelData variable were null when getting to line 237 which causes a null object reference exception on line 256. Have you seen this? I didn't know if maybe yall handled this exception higher up and I just missed it or if maybe I somehow am introducing this error, although it doesn't seem like I am.

**Comment by mia-san** [comment]
_pixels looks to be a (over)optimization for rendering last frame again and again.
by removing the 'optimization', we could eliminate _pixels variable, which would fix this issue altogether.
just an idea.

**Comment by CDarbonne** [comment]
I'm looking through it, but I'm not sure that I am grasping what you mean. What are you proposing should be removed?

**Comment by gofal** [comment]
The idea of DicomImage is to keep the raw pixels in memory and render just what you need as Bitmap/Bitmapsource. As a radiologist you do 2 kind of things:
* scrolling: you change the current index and draw one frame after the other
* windowzing/zooming/... then you see a interesting image and then you do windowing, zooming, panning etc. And there you have to render the same frame again and again with different parameters. This second case is what the _pixel is for. Here you can save lot of time if you don't have to extract the frame from the raw pixels again and again.

Of course this approach is not fine in terms of multithreading. If redesigning DicomImage then maybe we should create a DicomFrame class that holds the _pixels and some other parameters like PatientPosition or similar and the DicomImage holds a list of DicomFrames. They could be lazy loaded meaning instaniated on first access and then cached in the list? This could be done threadsafe easily, but this also means that you need double memory.

**Comment by CDarbonne** [comment]
@gofal I like this idea a lot. We are using multithreading a lot and would love to work with yall on figuring out a better, more efficient way of handling things like windowing, which currently requires quite a bit of overhead since `RenderImage()` has to be repeatedly called as the WindowCenter and WindowWidth values slide. I was about to create a custom windowing function to handle this, but I think your idea of having DicomFrames and caching in DicomImage is a superior route. Let us know how we can help with the next steps in your idea to this make this happen.

**Comment by mia-san** [comment]
@CDarbonne 
_pixels is used to cache decoded pixel data.
next invocation requests last frame again, _pixels is reused.
but _pixels is already known to be not 'threadsafe'.

however, _pixels is used for caching purpose only.
I thought that decoding pixel data every invocation and using local variable, _pixels could be eliminated.

of course this is just a bandaid.
@gofal 's idea is way to go for longer term.

**Comment by CDarbonne** [comment]
ahhh yes, thank you! I see what you mean now @mia-san! I made the changes and it no longer has that issue! I had also implemented a try-catch block that would recall renderImage recursively (as a bandaid), but now it never fires, so I think that fixed the root of the problem. 

I do agree that @gofal 's idea would be a better long-term solution/enhancement though, especially for windowing because, again, it has a lot of overhead in its current state and makes the visual not smooth at all.
